src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/MapCode/MapCodeEndpoint.cs

```diff
@@ -0,0 +1,440 @@
+                }
+
+                // We create a new Razor file based on each mapping's content in order to get the syntax tree that we'll later use to map.
+                var sourceDocument = RazorSourceDocument.Create(content, "Test" + extension);
+                var codeToMap = projectEngine.Process(sourceDocument, fileKind, importSources, tagHelperContext.TagHelpers);
+
+                await MapCodeAsync(codeToMap, mapping.FocusLocations, changes, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        var finalizedChanges = new Dictionary<string, LSP.TextEdit[]>();
+        foreach (var change in changes)
+        {
+            finalizedChanges.Add(change.Key, [.. change.Value]);
+        }
+
+        var workspaceEdits = new LSP.WorkspaceEdit
+        {
+            Changes = finalizedChanges,
+
+            // There's currently a bug on the client side that forces us to always return a non-null DocumentChanges.
+            DocumentChanges = Array.Empty<LSP.TextDocumentEdit>()
+        };
+
+        return workspaceEdits;
+    }
+
+    private async Task MapCodeAsync(
+        RazorCodeDocument codeToMap,
+        LSP.Location[][] locations,
+        Dictionary<string, List<LSP.TextEdit>> changes,
+        CancellationToken cancellationToken)
+    {
+        var syntaxTree = codeToMap.GetSyntaxTree();
+        if (syntaxTree is null)
+        {
+            return;
+        }
+
+        var nodesToMap = ExtractValidNodesToMap(syntaxTree.Root);
+        if (nodesToMap.Count == 0)
+        {
+            return;
+        }
+
+        await MapCodeAsync(locations, [.. nodesToMap], changes, cancellationToken).ConfigureAwait(false);
+        MergeEdits(changes);
+    }
+
+    private async Task MapCodeAsync(
+        LSP.Location[][] focusLocations,
+        SyntaxNode[] nodesToMap,
+        Dictionary<string, List<LSP.TextEdit>> changes,
+        CancellationToken cancellationToken)
+    {
+        var didCalculateCSharpFocusLocations = false;
+        var csharpFocusLocations = new LSP.Location[focusLocations.Length][];
+
+        // We attempt to map the code using each focus location in order of priority.
+        // The outer array is an ordered priority list (from highest to lowest priority),
+        // and the inner array is a list of locations that have the same priority.
+        // If we can successfully map using the first location, we'll stop and return.
+        foreach (var locationByPriority in focusLocations)
+        {
+            foreach (var location in locationByPriority)
+            {
+                // The current assumption is that all focus locations will always be in the same document
+                // as the code to map. The client is currently implemented using this behavior, but if it
+                // ever changes, we'll need to update this code to account for it.
+                var documentContext = _documentContextFactory.TryCreateForOpenDocument(location.Uri);
```

Comment:
Yep, I guess technically for now we can do the calculation outside of the loop until the client behavior changes (just updated).

> It certainly seems like the calling code only processes a new document for one piece of mapping.Contents at a time, so its not clear to me how it could apply to multiple

Focus locations are things like current cursor location, current visible scope, etc. They're organized in priority order based on which focus locations the model deems the most relevant. I believe the idea is that a focus location could differ from the file the model thinks we should insert into (e.g., user invokes AI in one file, but the model determines that the code should be inserted into a different file). Currently this scenario is unsupported by the client, but it's the reason why the spec separates focus location URIs from mapping content URIs.